Extended HI followup: fix speed issue

yujiex opened the pull request:
Pull request overview
---------------------
 - This PR fixes the speed problem of root solver in extended HI calculation. The original PR for extended heat index is here: https://github.com/NREL/EnergyPlus/pull/10548
 
A couple of different approaches were tried out to improve the performance

- changing the root solver type from `Bisection` to `BisectionThenRegulaFalsi`: this had the biggest impact, bringing the damage from 70% to about 12% 
- result caching: saving Zs(Rs) result to a variable before entering the root solvers. This improved about 1-2 percentage point. 
- limit the domain of application to the region where the two methods differ more (red and blue region). That doesn't help much either (< 0.5% improvement). 
 
The new extend HI method might just requires a lot of computation. So The following object is added to provide a switch from the simplified to the extended HI method.

```
OutputControl:ResilienceSummaries,
       \memo Specifies methods for resilience reporting variables
       \unique-object
  A1 ; \field Heat Index Algorithm
       \note Whether the simplified or the extended method should be used for heat index
       \note Simplified: based on regression analysis carried out by Lans P. Rothfusz
       \note Extended: Based on paper by Lu & Romps
       \type choice
       \key Simplified
       \key Extended
       \default Simplified
```

#### Regression diff

There's diff in one file: 1ZoneUncontrolled_Win_ASH55_Thermal_Comfort.idf. The diff is a result of the added OutputControl:ResilienceSummaries
![image](https://github.com/user-attachments/assets/eac22aca-b47d-4914-bdaf-dc6646f8df58)


**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
change rootsolver to BisectionThenRegulaFalsi to speed up

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHIspeedFix

Commit:
Revert "Revert "Adopt the Extended Heat Index calculation in zone resilience""

This reverts commit a0f92d6fdf8a9666076dee4c916113d42225b6d0.

Commit:
change root solver to BisectionThenRegulaFalsi for speed

Commit:
save Zs(Rs) value in a variable

Commit:
find eqvar name and value together

Commit:
update unit test find_eqvar_name and value

Commit:
save Ra_bar calculation result

yujiex committed:
restrict domain of extended HI application

yujiex committed:
Revert "restrict domain of extended HI application"

This reverts commit 58a77d92d15c64d0e2917ef546ba3336eda4cd1b.

yujiex committed:
try using a shorter termination for bisection solver

Myoldmopar commented:
Even with the latest commit, [4f4809f](https://github.com/NREL/EnergyPlus/pull/10733/commits/4f4809f4777513d8db5240bf383a138c2df38612), it's still causing a [10% increase](https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-4f4809f4777513d8db5240bf383a138c2df38612-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html) in instruction count in just our performance test files.  We can figure out a solution for this, but it doesn't look like it's going in this release.  I'm marking this for next release and we'll deal with this later.  

If this method is trimmed as far down as possible but still this much slower, I would suggest adding a switch to calculate heat index using the old way by default and then enabling the new method as an alternative.  This would be a user-facing input option, so definitely not in this current release.

yujiex commented:
Thanks, @Myoldmopar . It seems that changing the solver type had the biggest impact. Then I tried adding some result caching but those doesn't save as much (like improved 1-2 percentage point). I also tried to limit the domain of its application (in the old PR there was a plot of how results differ between the old and the new methods. I restricted it to only apply on the bottom right region where the two methods differ a lot). That doesn't help much either. Potentially the new method just requires a lot of computation.

Commit:
Merge remote-tracking branch 'origin/extendedHIspeedFix' into extendedHIspeedFix

Commit:
Try using RegulaFalsiThenBisection

Commit:
Revert "save Ra_bar calculation result"

This reverts commit 660006965463ee23abbe207c44d4a03b6471d77f.

Commit:
Revert "Try using RegulaFalsiThenBisection"

This reverts commit ec8f0babb51302212ceeed61cdeed0ca33db43f4.

Commit:
Revert "Revert "restrict domain of extended HI application""

This reverts commit 16faa3137327f745939434b8b68750da66aebb8a.

Commit:
fix application domain equation

Commit:
Add global control of heat index method choice

Commit:
fix unit test

yujiex commented:
@Myoldmopar I added an idd object to switch heat index calculation method. Would you mind taking a look? Thanks :)

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHIspeedFix

nrel-bot-2 commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 17 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 8 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

Myoldmopar commented on doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex:
```diff
@@ -66,69 +66,17 @@ \subsubsection{Heat Index}\label{heat-index}
 \end{table}
 
-The computation of the heat index is a refinement of a result obtained by
+Before version 24.2, the computation of the heat index is a refinement of a result obtained by
```
Need to bump this to 25.1 if this is going in now.

yujiex replied:
good catch! just changed the version number.

Myoldmopar commented on doc/engineering-reference/src/special-modules-reporting/resilience-metrics.tex:
```diff
@@ -66,69 +66,17 @@ \subsubsection{Heat Index}\label{heat-index}
-calculation is based on degree Fahrenheit.
+1990 National Weather Service (NWS) Technical Attachment (SR 90-23) [4-5].
 
-The regression equation of Rothfusz is
```
Did this stuff just get removed?

yujiex replied:
yeah... I reverted the change of the previous extended heat index commit and at the time that one is supposed to replace the Rothfusz-Steadman used before. I forgot to put these documentation back. Just added them back.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -977,6 +977,17 @@ HVACSystemRootFindingAlgorithm,
+       \type choice
+       \key Simplified
+       \key Extended
+       \default Simplified
```
I'm glad there is a switch in place now.  @mjwitte do you feel like there's any better spot for this switch?  Seeing a new root-level object was a surprise, but maybe it's indeed the best fit.

mjwitte replied:
If it stays as a separate object, I might suggest something like:
```
OutputControl:ResilienceSummaries,
\memo . . . 
\unique-object
  A1 ; \field Heat Index Algorithm
      \note ....
       \type choice
       \key Simplified
       \key Extended
       \default Simplified
```
This would allow for additional options in the future for thermal or other resilience reports.

OR, since this is related to thermal comfort, would it make sense to add this as a field in the People object? Then this could be set to different values for different People objects if desired. It appears there are already some fields here that relate to thermal resilience? One negative point would be that it requires changing every People object in order to use the Extended calculation everywhere.

```
People,

  N7,  \field Cold Stress Temperature Threshold
       \type real
       \units C
       \note this is the indoor safe temperature threshold for cold stress
       \default 15.56
  N8,  \field Heat Stress Temperature Threshold
       \type real
       \units C
       \note this is the indoor safe temperature threshold for heat stress
       \default 30
  A22 ; \field Heat Index Algorithm
      \note ....
       \type choice
       \key Simplified
       \key Extended
       \default Simplified
```

No matter where this lands it will need I/O Ref documentation.

yujiex replied:
Currently the heat index calculation is at zone level
Should I move it to people level if we want to use the people option?
```
if (state.dataHeatBalSurfMgr->reportVarHeatIndex || state.dataOutRptTab->displayThermalResilienceSummary) {
    for (int ZoneNum = 1; ZoneNum <= state.dataGlobal->NumOfZones; ++ZoneNum) {
        Real64 const ZoneT = state.dataZoneTempPredictorCorrector->zoneHeatBalance(ZoneNum).ZTAV;
        Real64 const ZoneW = state.dataZoneTempPredictorCorrector->zoneHeatBalance(ZoneNum).airHumRatAvg;
        Real64 const ZoneRH = Psychrometrics::PsyRhFnTdbWPb(state, ZoneT, ZoneW, state.dataEnvrn->OutBaroPress) * 100.0;
        Real64 const ZoneTF = ZoneT * (9.0 / 5.0) + 32.0;
        // heat index calculation here
    }
}
```
I'll add in the IO reference documentation.

mjwitte replied:
> Should I move it to people level if we want to use the people option?

No. I didn't look closely enough at that. Are you (@yujiex and @Myoldmopar) ok with the `OutputControl:ResilienceSummaries` object?

yujiex replied:
The `OutputControl:ResilienceSummaries` looks good

yujiex replied:
I just changed the idd object to OutputControl:ResilienceSummaries @mjwitte

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -262,6 +262,8 @@ set(SRC
     ExhaustAirSystemManager.cc
     ExhaustAirSystemManager.hh
+    ExtendedHI.cc
+    ExtendedHI.hh
```
This is random, but since I think there will be some changes needed, could we just rename this ExtendedHeatIndex?  Or could we pull the base heat index calculations into this file too and call it HeatIndex.cc?  Minor change, and I can make this change during my review, you don't have to.

yujiex replied:
Just changed the names to ExtendedHeatIndex

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,501 @@
+// EnergyPlus, Copyright (c) 1996-2024, The Board of Trustees of the University of Illinois,
```
The copyrights in the new files will need to be bumped to 2025.

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,501 @@
+    constexpr Real64 cpv = cvv + rgasv;
+
+    // The saturation vapor pressure
+    Real64 pvstar(Real64 T)
```
Random, but `const T`

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,501 @@
+    constexpr Real64 Pa0 = 1.6e3;                     // Pa        , reference air vapor pressure in regions III, IV, V, VI, steadman1979
+
+    // Function to calculate respiratory heat loss, W/m^2
+    Real64 Qv(Real64 Ta, Real64 Pa)
```
Random, but it would be good to apply `const` wherever possible in these functions (arguments and locals).

yujiex replied:
make sense. Just changed the arguments to const ones.

Myoldmopar commented on src/EnergyPlus/ExtendedHI.cc:
```diff
@@ -0,0 +1,501 @@
+        // RH: relative humidity in range of 0.0 to 1.0
+        // The function computes the extended heat index, in Kelvinn
+
+        auto const HVACSystemRootSolverBackup = state.dataRootFinder->HVACSystemRootFinding.HVACSystemRootSolver;
```
I know this caught my attention before because I was thinking you were making a copy of a whole solver object.  It would be great to rename this to like "RootSolverMethod" or something to make clear it's just a flag.

yujiex replied:
yeah, it does sounds like a large object. I just renamed it to HVACSystemRootSolverMethod

Myoldmopar commented on src/EnergyPlus/ExtendedHI.hh:
```diff
@@ -0,0 +1,90 @@
+// EnergyPlus, Copyright (c) 1996-2024, The Board of Trustees of the University of Illinois,
```
2025

yujiex replied:
updated them

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,628 @@
+// EnergyPlus, Copyright (c) 1996-2024, The Board of Trustees of the University of Illinois,
```
2025

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,628 @@
+    return HI;
+}
+
+constexpr Real64 Fahrenheit2Celsius(Real64 F)
```
I think there must be a function inside E+ code itself rather than another copy of it.

yujiex replied:
I see. There's a TempIPtoSI function.

Myoldmopar reviewed (commented):
OK, definitely needs some changes before moving forward, although they might be minor. Let me know your thoughts on a few of the comments and we'll see how to nudge this back along.

Commit:
add back doc about Rothfusz-Steadman method, change version num

Commit:
change function args to const arg

Commit:
change local var to const

Commit:
rename HVACSystemRootSolver to HVACSystemRootSolverMethod

Commit:
use existing F to C conversion function

Commit:
Merge remote-tracking branch 'origin/develop' into extendedHIspeedFix

Commit:
license year update

Commit:
File rename ExtendedHI to ExtendedHeatIndex

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -1217,17 +1219,17 @@ namespace HeatBalanceManager {
                 {
                     std::string const &SELECT_CASE_var = AlphaName(1);
```
This is what `getEnumValue` is for.

yujiex replied:
yeah, let me change this to use getEnumValue

amirroth commented on src/EnergyPlus/ExtendedHeatIndex.cc:
```diff
@@ -0,0 +1,501 @@
+    constexpr Real64 epsilon = 0.97;                  //           , emissivity of surface, steadman1979
+    constexpr Real64 M = 83.6;                        // kg        , mass of average US adults, fryar2018
+    constexpr Real64 H = 1.69;                        // m         , height of average US adults, fryar2018
+    Real64 A = 0.202 * pow(M, 0.425) * pow(H, 0.725); // m^2       , DuBois formula, parson2014
```
Don't use global variables.

Commit:
Change idd obj name to OutputControl:ResilienceSummaries w doc change

also moved heat index calculation into simulation manager after the other
OutputControl thing

Commit:
use getEnumValue for HVACSystemRootSolverMethod

Commit:
remove global variables

Commit:
clang-format

Commit:
license year in .hh

amirroth commented on src/EnergyPlus/ExtendedHeatIndex.cc:
```diff
@@ -0,0 +1,552 @@
+        Real64 Rf;
+
+        if (flux1 <= 0.0) {
+            varname = static_cast<int>(EqvarName::Phi);
```
Why cast this to int all the time? Just use the enum, no?

yujiex replied:
Yeah I should use the enum class instead of casting to int

amirroth commented on src/EnergyPlus/ExtendedHeatIndex.cc:
```diff
@@ -0,0 +1,552 @@
+    }
+
+    // Convert the find_T function
+    Real64 find_T(EnergyPlusData &state, int const eqvar_name, Real64 const eqvar)
```
You can use `enum`s as parameters, why make it an `int` here?

yujiex replied:
yeah also need to use enum class directly here as well

amirroth commented on src/EnergyPlus/ExtendedHeatIndex.hh:
```diff
@@ -0,0 +1,90 @@
+    enum class EqvarName
+    {
+        Invalid = -1,
+        Phi = 1,
```
Why are these not starting at 0? Why do you have to assign explicit values to them?

yujiex replied:
This extended heat index was based on python code on [this page](https://romps.berkeley.edu/papers/pubs-2020-heatindex.html). They have some dictionary defined the indices like this: `dic = {"phi":1,"Rf":2,"Rs":3,"Rs*":3,"dTcdt":4}` So I just used the same indices.

amirroth replied:
The values are meaningless unless you are using them in an array, but even then you should use them in a `std::array` and they should start at 0.

yujiex replied:
make sense. Let me change this to starting at 0.

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -1214,36 +1214,25 @@ namespace HeatBalanceManager {
+                    static_cast<HVACSystemRootSolverAlgorithm>(getEnumValue(HVACSystemRootSolverAlgorithmUC, Util::makeUPPER(AlphaName(1))));
+                if (HVACSystemRootFinding.HVACSystemRootSolverMethod == HVACSystemRootSolverAlgorithm::Invalid) {
+                    HVACSystemRootFinding.HVACSystemRootSolverMethod = HVACSystemRootSolverAlgorithm::RegulaFalsi;
+                    ShowWarningError(state,
```
There is a shortcut for this warning called `ShowWarningInvalidKey`.

yujiex replied:
Oh! Didn't realize there's this shortcut.

amirroth commented on src/EnergyPlus/ExtendedHeatIndex.cc:
```diff
@@ -0,0 +1,552 @@
+    constexpr Real64 Za_bar = 60.6 / 11.6; // Pa m^2/W, mass transfer resistance through air, clothed part of skin
+    constexpr Real64 Za_un = 60.6 / 12.3;  // Pa m^2/W, mass transfer resistance through air, when being naked
+
+    constexpr Real64 tol = 1e-8;
```
This is an absolute tolerance level. What are the reasonable ranges of values here and what level of precision does this tolerance correspond to relative to those values?

yujiex replied:
The tolerance was from the python code here [this page](https://romps.berkeley.edu/papers/pubs-2020-heatindex.html).

amirroth replied:
What someone else chose for the tolerance is not relevant. What is the magnitude of the values produced by this function? 1e-4? 1e-1? 1e+2?

amirroth replied:
If this function produces SGP (whatever that is) and the values are on the order of 1e-2 then a tolerance of 1e-8 is about six decimal places of precision which is about twice what is needed. I cannot conceive of any quantity that EnergyPlus that would need to know to a precision of more than 0.1% (1% may be okay for many thing too).  

In fact, `SolveRoot` should probably be revised to use relative tolerance after some small fixed number of iterations. I wonder how much performance over-specific root precision is costing us. It could be quite a lot.

yujiex replied:
If I change tol to 1e-4, there will be a 0.6 difference for a target value of about 360 (0.2% relative difference). Using tol=1e-5 keeps the difference to below 0.1 (for values ranging from 200 to 400). Maybe I'll change it to tol = 1e-5?

amirroth replied:
Absolutely. Those additional digits are costing you iterations for no real benefit in terms of accuracy. Who cares if the SGP is "off" by less than 0.1%? How accurate is that metric anyway?

yujiex replied:
Normally heat index is reported at either integer or 1 decimal point. The range is mostly two three hundred. Hopefully within 0.1C accuracy is good enough.

amirroth replied:
Most thermostats used in buildings do not have this resolution.

yujiex replied:
yeah the inputs of temperature and relative humidity percentage points are integers

Commit:
directly use the enum class EqvarName instead of casting

Commit:
use ShowWarningInvalidKey for warning message

Commit:
more removal of global vars

Commit:
clang-format, change enum EqvarName to start from 0

Commit:
change root finding tolerance, clang-format

mjwitte commented on doc/input-output-reference/src/input-for-output.tex:
```diff
@@ -346,6 +346,21 @@ \subsubsection{Inputs}\label{inputs-7-021}
+
+This input specifies methods or algorithms used in certain resilience reporting variables.
+
+\subsubsection{Inputs}\label{inputs-8-020}
```
When viewing the changes on github, there is a warning about this line:
`Label `inputs-8-020' multiply defined.`

yujiex replied:
let me check on why this happens

yujiex replied:
I see, my label for the inputs section in my latex documentation collides with a existing one. I just changed the label.

mjwitte reviewed (commented):
@yujiex Thanks for the idd changes.

Commit:
doc input section latex label fix

Myoldmopar commented on src/EnergyPlus/ExtendedHeatIndex.cc:
```diff
@@ -0,0 +1,560 @@
+    }
+
+} // namespace ExtendedHI
+} // namespace EnergyPlus
```
The strong use of const and constexpr in this file make me very happy.

Myoldmopar commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -1013,6 +1013,27 @@ namespace SimulationManager {
             }
         }
 
+        state.dataHeatBalMgr->CurrentModuleObject = "OutputControl:ResilienceSummaries";
```
In the future, there's no need to use a state variable to hold this string.  Just keep a constexpr string view variable of it and use that.

yujiex replied:
I'll remember and use string view in the future.

Myoldmopar commented on tst/EnergyPlus/unit/ExtendedHI.unit.cc:
```diff
@@ -0,0 +1,589 @@
+TEST_F(EnergyPlusFixture, extendedHI_pvstar)
+{
+    Real64 tol = 1e-8;
+    std::vector<Real64> T_values = {200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370};
```
You could use `constexpr std::array` all over the place in here.  Since this is just in the tests, we don't have to hold this up, but always look for that in the future.

yujiex replied:
oh! I see I see. I'll take a note and remember to use constexpr next time.

Myoldmopar reviewed (approved):
Outside of the vector/array thing in the unit test code, I don't have any comments.  And that's not worth holding this up.  I will do final testing and try to get this in.

Myoldmopar commented:
Runs great here, ready to go in.  Thanks @yujiex.
